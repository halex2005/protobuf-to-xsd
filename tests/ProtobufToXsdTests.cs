using System;
using System.Diagnostics;
using System.IO;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Reporters;
using NUnit.Framework;

namespace tests
{
	[UseReporter(typeof(DiffReporter))]
	[TestFixture]
	public class ProtobufToXsdTests
	{
		private string ConvertProtobufToXsd(string protoFilePath)
		{
			var currentAssemblyDirectory = Path.GetDirectoryName(new Uri(typeof(ProtobufToXsdTests).Assembly.CodeBase).AbsolutePath);
			var tempFile = Path.GetTempFileName();
			try
			{
				var protogenPath = Path.Combine(
					currentAssemblyDirectory,
					"protobuf-to-xsd",
					"protogen.exe");
				var arguments = $"-i:\"{protoFilePath}\" -o:\"{tempFile}\" -d -t:xsd-attributes";
				Console.WriteLine($"Starting process: \"{protogenPath}\" {arguments}");

				var startInfo = new ProcessStartInfo(protogenPath, arguments)
				{
					CreateNoWindow = true,
					UseShellExecute = false,
					WorkingDirectory = currentAssemblyDirectory,
					RedirectStandardError = true,
					RedirectStandardOutput = true,
				};

				var process = Process.Start(startInfo);
				process.BeginErrorReadLine();
				process.BeginOutputReadLine();
				process.OutputDataReceived += (s, o) => { Console.Write(o.Data); };
				process.ErrorDataReceived += (s, o) => { Console.Write(o.Data); };
				process.WaitForExit();
				Console.WriteLine($"Exit code: {process.ExitCode}");
				return File.ReadAllText(tempFile);
			}
			finally
			{
				File.Delete(tempFile);
			}
		}

		[Test]
		public void FieldsConvertingTests()
		{
			var result = ConvertProtobufToXsd("protobuf-fields.proto");
			Approvals.VerifyXml(result);
		}
	}
}
