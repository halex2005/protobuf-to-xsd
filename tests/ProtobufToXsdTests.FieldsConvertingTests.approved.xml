<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:msxsl="urn:schemas-microsoft-com:xslt">
  <!--Generated from: protobuf-fields.proto-->
  <!--Namespace: protobuf-fields-->
  <xs:complexType name="RootElement">
    <xs:all>
      <xs:element name="ComplexRequired" type="ChildElement" />
      <xs:element name="ComplexOptional" type="ChildElement" minOccurs="0" />
      <xs:element name="ComplexRepeateds">
        <xs:complexType>
          <xs:sequence>
            <xs:element name="ComplexRepeated" type="ChildElement" maxOccurs="unbounded" />
          </xs:sequence>
        </xs:complexType>
      </xs:element>
      <xs:element name="SimpleRepeateds">
        <xs:complexType>
          <xs:sequence>
            <xs:element name="SimpleRepeated" type="xs:string" />
          </xs:sequence>
        </xs:complexType>
      </xs:element>
      <xs:element name="EnumeratedRepeateds">
        <xs:complexType>
          <xs:sequence>
            <xs:element name="EnumeratedRepeated" type="EnumerationElement" />
          </xs:sequence>
        </xs:complexType>
      </xs:element>
    </xs:all>
    <xs:attribute name="SimpleRequired" use="required" type="xs:string" />
    <xs:attribute name="SimpleRequiredWithDefault" use="required" default="SimpleRequiredDefaultValue" type="xs:string" />
    <xs:attribute name="EnumeratedRequired" use="required" type="EnumerationElement" />
    <xs:attribute name="EnumeratedRequiredWithDefault" use="required" type="EnumerationElement" />
    <xs:attribute name="SimpleOptional" use="optional" type="xs:string" />
    <xs:attribute name="SimpleOptionalWithDefault" use="optional" default="SimpleOptionalDefaultValue" type="xs:string" />
    <xs:attribute name="EnumeratedOptional" use="optional" type="EnumerationElement" />
    <xs:attribute name="EnumeratedOptionalWithDefault" use="optional" type="EnumerationElement" />
  </xs:complexType>
  <xs:complexType name="ChildElement">
    <xs:all />
    <xs:attribute name="ChildValue" use="required" type="xs:string" />
  </xs:complexType>
  <xs:simpleType name="EnumerationElement">
    <xs:restriction base="xs:string">
      <!--FirstValue:-->
      <xs:enumeration value="1" />
      <!--SecondValue:-->
      <xs:enumeration value="2" />
    </xs:restriction>
  </xs:simpleType>
</xs:schema>